controllers/operatorpolicy_status.go

```diff
@@ -0,0 +1,429 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	policyv1 "open-cluster-management.io/config-policy-controller/api/v1"
+	policyv1beta1 "open-cluster-management.io/config-policy-controller/api/v1beta1"
+)
+
+// updateStatus takes one condition to update, and related objects for that condition. The related
+// objects given will replace all existing relatedObjects with the same gvk. If a condition is
+// changed, the compliance will be recalculated and a compliance event will be emitted. The
+// condition and related objects can match what is already in the status - in that case, no API
+// calls are made. The `lastTransitionTime` on a condition is not considered when checking if the
+// condition has changed. If not provided, the `lastTransitionTime` will use "now". It also handles
+// preserving the `CreatedByPolicy` property on relatedObjects.
+//
+// This function assumes that all given related objects are of the same kind.
+//
+// Note that only changing the related objects will not emit a new compliance event, but will update
+// the status.
+func (r *OperatorPolicyReconciler) updateStatus(
```

Comment:
I think the controller should send one compliance event per compliance change like the config-policy-controller to prevent so many compliance changes on the policy.

The config-policy-controller does it with "event batches" at the end of `handleObjectTemplates`. I imagine it'd be a lot simpler in the operator policy case since there is no namespace selector and not multiple object templates.

Reply:
The number of status updates and compliance events is a valid critique. I wanted to send the updates and event immediately, because I felt it simplified error handling and limited the amount of information that needed to be returned by other functions.

I find what the config policy does with event batches very difficult to reason through... Can you carefully define "one compliance event per compliance change"?

Imagine a fresh enforced policy. With this implementation we get 4 status updates and events that look kinda like this:
- NonCompliant; the operator group is missing (subscription details omitted because they are still unknown)
- NonCompliant; the operator group was created (subscription details omitted because they are still unknown)
- NonCompilant; the operator group was created and the subscription is missing
- Compliant; the operator group was created and the subscription was created

Are you asking for just two events like these?
- NonCompliant; the operator group is missing, and the subscription is missing, 
- Compliant; the operator group was created, and the subscription was created

Reply:
Yeah, exactly. If I have time today, I can try to make the event batching logic more generic since I wrote that code.

The gist of the approach is that each `handle` method would return a list of events. So say operator group returned three events (two noncompliant) and subscription returned two event (one noncompliant), you would end up with three events in this way:

1. noncompliant operator group + existing state
1. noncompliant operator group and noncompliant subscription
1. compliant operator group and compliant subscription